File: app-ios/VelaWallet/VelaWallet/Components/ActionStack.swift
//
//  ActionStack.swift
//  VelaWallet
//
//  The single authoritative outcome action stack (spec 014): one primary
//  CTA (the existing VelaButton) followed by up to two secondaries styled
//  as the mock's dark solid rows — NOT the outlined welcome-secondary
//  style (contract §5). Presses emit ActionIds only; the host decides.
//

import SwiftUI

/// One resolved, renderable action: components take resolved strings.
struct ActionEntry: Identifiable {
    let id: ActionId
    let role: ActionRole
    let title: String
}

struct ActionStack: View {
    let actions: [ActionEntry]
    let onAction: (ActionId) -> Void

    var body: some View {
        VStack(spacing: FlowGeometry.actionGap) {
            ForEach(actions) { entry in
                switch entry.role {
                case .primary:
                    VelaButton(title: entry.title, kind: .primary) {
                        onAction(entry.id)
                    }
                case .secondary:
                    SecondaryRowButton(title: entry.title) {
                        onAction(entry.id)
                    }
                }
            }
        }
    }
}

/// The mock's dark filled secondary row: full width, control.lg height,
/// sunken fill with a hairline for light-theme definition.
private struct SecondaryRowButton: View {
    let title: String
    let action: () -> Void

    var body: some View {
        Button(action: action) {
            Text(title).typeRole(Typography.button)
        }
        .buttonStyle(SecondaryRowStyle())
    }
}

private struct SecondaryRowStyle: ButtonStyle {
    @Environment(\.theme) private var theme

    func makeBody(configuration: Configuration) -> some View {
        let pressed = configuration.isPressed
        // minHeight, not a fixed height: long-locale labels wrap centered and
        // grow the row instead of being clipped (spec 014 long-label fix).
        return configuration.label
            .foregroundStyle(theme.fgBase)
            .multilineTextAlignment(.center)
            .padding(.horizontal, Tokens.Space.s24)
            .padding(.vertical, Tokens.Space.s8)
            .frame(maxWidth: .infinity)
            .frame(minHeight: FlowGeometry.actionRowHeight)
            .background {
                RoundedRectangle(cornerRadius: Tokens.Radius.r16)
                    .fill(theme.bgSunken)
                RoundedRectangle(cornerRadius: Tokens.Radius.r16)
                    .strokeBorder(theme.borderBase, lineWidth: Tokens.BorderWidth.hairline)
            }
            .opacity(pressed ? Interaction.pressedOpacity : 1)
            .animation(.easeOut(duration: Tokens.Motion.fast), value: pressed)
    }
}

#Preview("Action stack") {
    ActionStack(
        actions: [
            ActionEntry(id: .retry, role: .primary, title: "重试"),
            ActionEntry(id: .editIndexEndpoint, role: .secondary, title: "修改索引服务地址"),
            ActionEntry(id: .reportError, role: .secondary, title: "上报这个错误"),
        ],
        onAction: { _ in }
    )
    .padding(Tokens.Space.s24)
    .themed(.light)
}

#Preview("Action stack dark") {
    ActionStack(
        actions: [
            ActionEntry(id: .retry, role: .primary, title: "重试"),
            ActionEntry(id: .editIndexEndpoint, role: .secondary, title: "修改索引服务地址"),
            ActionEntry(id: .reportError, role: .secondary, title: "上报这个错误"),
        ],
        onAction: { _ in }
    )
    .padding(Tokens.Space.s24)
    .background(Tokens.dark.bgRaised.color)
    .themed(.dark)
}

File: app-desktop/vela-wallet/src/ui/action_stack.rs
//! The outcome pattern's stacked actions — one authority for the primary
//! capsule + dark solid secondary rows treatment (spec 014 contract §5).
//! Presses emit the action's [`ActionId`] to the host sink; the stack never
//! decides what happens next.

use crate::onboarding_flow::{Action, ActionId, ActionRole};
use crate::theme::{FLOW_GAP_MD, Theme};
use crate::ui::{ButtonVariant, vela_button_opts};
use gpui::{App, Div, ParentElement, SharedString, Styled, Window, div, px};

pub fn action_stack(
    theme: &Theme,
    actions: &[Action],
    on_action: impl Fn(ActionId, &mut Window, &mut App) + Clone + 'static,
) -> Div {
    let mut col = div().flex().flex_col().gap(px(FLOW_GAP_MD)).w_full();
    for (ix, action) in actions.iter().enumerate() {
        let variant = match action.role {
            ActionRole::Primary => ButtonVariant::Primary,
            // NOT the outlined welcome secondary: the mock's solid rows.
            ActionRole::Secondary => ButtonVariant::Row,
        };
        let id = action.id;
        let press = on_action.clone();
        col = col.child(vela_button_opts(
            SharedString::from(format!("flow-action-{ix}")),
            variant,
            action.label.clone(),
            true,
            theme,
            move |_, window, cx| press(id, window, cx),
        ));
    }
    col
}
